src/app/plugins/ta.plugin.ts

```diff
@@ -0,0 +1,153 @@
+import { Plugin } from '../../common/plugin';
+import { IContainer, IMessage, ChannelType, Maybe } from '../../common/types';
+import { GuildChannel, GuildMember, TextChannel, Role } from 'discord.js';
+
+export class TaPlugin extends Plugin {
+  public name: string = 'TA Plugin';
+  public description: string = 'Allows TAs to register for classes.';
+  public usage: string = 'ta register <class_name> | ta ask <question>';
+  public pluginAlias = [];
+  public permission: ChannelType = ChannelType.Private;
+
+  private _TA_ROLE = 'Teaching Assistant';
+  private _roleCache: Record<string, Maybe<Role>> = {
+    [this._TA_ROLE]: undefined,
+  };
+
+  constructor(public container: IContainer) {
+    super();
+  }
+
+  public validate(message: IMessage, args: string[]) {
+    return !!args.length;
+  }
+
+  public async execute(message: IMessage, args?: string[]) {
+    if (!args?.length) {
+      return;
+    }
+
+    if (!this._roleCache[this._TA_ROLE]) {
+      this._roleCache[this._TA_ROLE] = this.container.guildService
+        .get()
+        .roles.filter((r) => r.name === this._TA_ROLE)
+        .first();
+    }
+
+    const member = this.container.guildService.get().members.get(message.author.id);
+    if (!member) {
+      return;
+    }
+
+    const [subcommand, ...question] = args;
+    if (subcommand === 'ask') {
+      this._handleAsk(message, question.join(' '));
+      return;
+    }
+
+    const hasRole = member.roles.array().some((r) => r === this._roleCache[this._TA_ROLE]);
+    if (!hasRole) {
+      message.reply('You are not a TA!');
+      return;
+    }
+
+    if (subcommand === 'register') {
+      this._handleRegister(message);
+      return;
+    }
+
+    if (subcommand === 'remove') {
+      this._handleRemove(message);
+      return;
+    }
+
+    message.reply('Invalid sub-command');
+  }
+
+  private _handleAsk(message: IMessage, question: string) {
+    const channelTopic = (<TextChannel>message.channel).topic || '';
+    const hasTA: Boolean = channelTopic.indexOf('TA: ') !== -1;
+
+    if (!hasTA) {
+      message.reply('There are no TAs registered in this channel');
+      return;
+    }
+
+    const TA_tags = channelTopic.split('| TA: ')[1].split(' ');
```

Comment:
how does this work if someone has a space in their name?

Reply:
My bad, didn't know tags could have spaces